include/zephyr/drivers/comparator.h

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_COMPARATOR_H_
+#define ZEPHYR_INCLUDE_DRIVERS_COMPARATOR_H_
+
+/**
+ * @brief Comparator Driver APIs
+ * @defgroup comparator_interface Comparator Driver APIs
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/dt-bindings/comparator/comparator.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util_macro.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @name Comparator state flags
+ * @{
+ */
+
+/** Positive input voltage is below the negative input voltage. */
+#define COMPARATOR_STATE_BELOW BIT(0)
+/** Positive input voltage is above the negative input voltage. */
+#define COMPARATOR_STATE_ABOVE BIT(1)
+
+/** @cond INTERNAL_HIDDEN */
+#define COMPARATOR_STATE_RESULT_MASK  0x3UL
+/* Position in the state bitfield from which vendor-specific flags can be defined. */
+#define COMPARATOR_STATE_VENDOR_POS 8
+/** @endcond */
+
+/** @} */
+
+/**
+ * @brief Comparator callback.
+ *
+ * @param dev Comparator device instance.
+ * @param state Bitfield that specifies the reported state. This parameter can
+ *              contain COMPARATOR_STATE_* values and vendor-specific flags.
+ * @param user_data User data.
+ */
+typedef void (*comparator_callback_t)(const struct device *dev,
+				      uint32_t state,
+				      void *user_data);
+
+/** @brief Comparator configuration structure. */
+struct comparator_cfg {
+	/**
+	 * Positive input specification.
+	 * This field is composed of vendor-specific values.
+	 */
+	uint16_t input_positive;
+	/**
+	 * Negative input specification.
+	 * This field is composed of vendor-specific values.
+	 */
+	uint16_t input_negative;
+	/**
+	 * Flags with additional configuration options.
+	 * This field can contain COMPARATOR_FLAG_* values and vendor-specific
+	 * flags.
+	 */
+	uint32_t flags;
+};
```

Comment:
I think this would necessarily need to change in order to support comparator IP other than Nordic's.

E.g. https://mcuxpresso.nxp.com/api_doc/dev/1573/a00010.html#a00211

Additionally, some silicon might not even require a `flags` field and might simply store flags in a register.

It might be premature to declare `comparator_cfg` this way. Private flags like the sensor API.

Are there existing Linux DT bindings?